### PR TITLE
Detect public api changes

### PR DESCRIPTION
**Pull request by goergisn**
## Summary
- Adds a github action to detect public api changes

### Possible improvements
- [ ] Show a status/emoji that quickly represents whether or not there are actual changes
- [ ] (Speed) Optimization (See TODOs)

### CODE REVIEW

File: Scripts/compare_public_interface_definition.swift [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    
+    enum CodingKeys: String, CodingKey {
+        case printedName
+    }
+}
+
+class Element: Codable, Equatable, CustomDebugStringConvertible {
+    let kind: String
+    let name: String
+    let mangledName: String?
```
Comment (nauaros) [inline Scripts/compare_public_interface_definition.swift:28]: What's `mangledName`?

Comment (goergisn) [reply Scripts/compare_public_interface_definition.swift:28]: it's part of the sdk-dump json and looked like a more unique naming than name - could probably only rely on name but felt more unique 🤔 

Unfortunately I didn't find any documentation regarding the structure generated by the sdk dump option

Comment (nauaros) [reply Scripts/compare_public_interface_definition.swift:28]: Ah I see

File: Scripts/compare_public_interface_definition.swift [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        // TODO: Show what exactly changed (name, spi, conformance, declAttributes, ...) as a bullet list maybe (add a `changeList` property to `Change`)
+        changes += [.init(changeType: .change, parentName: lhs.parentPath, changeDescription: "`\(lhs)` ➡️  `\(rhs)`")]
+    }
+    
+    if lhs.children == rhs.children {
+        return changes
+    }
+    
+    changes += lhs.children?.flatMap { lhsElement in
+        if let rhsChildForName = rhs.children?.first(where: { ($0.mangledName ?? $0.name) == (lhsElement.mangledName ?? lhsElement.name) }) {
```
Comment (nauaros) [inline Scripts/compare_public_interface_definition.swift]: It would be nice to refactor this line. Not that easy to read. It can be a following iteration

**Review by nauaros — DISMISSED** [review]

File: Scripts/compare_public_interface_definition.swift (hunk 2 above)
Comment (goergisn) [reply Scripts/compare_public_interface_definition.swift]: Good one - I love extensions you know 💚

**Review by nauaros — DISMISSED** [review]

File: Scripts/generate_public_interface_definition.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+function setupComparisonRepo() {
+    rm -rf $COMPARISON_VERSION_DIR_NAME
+    mkdir $COMPARISON_VERSION_DIR_NAME
+    cd $COMPARISON_VERSION_DIR_NAME
+    git clone -b $BRANCH $REPO
+    
+    cd adyen-ios
+    mv Adyen.xcodeproj Adyen.xcode_proj
```
Comment (atmamont) [inline Scripts/generate_public_interface_definition.sh]: Let's add comment here for the reasoning?

**Review by atmamont — DISMISSED** [review]
👏 This will be truly helpful!

File: Scripts/generate_public_interface_definition.sh (hunk 3 above)
Comment (goergisn) [reply Scripts/generate_public_interface_definition.sh]: Good one - added 👌

**Review by atmamont — APPROVED** [review]

**Pull request merged**